packages/expect/src/toThrowMatchers.js

```diff
@@ -119,27 +136,65 @@ const toThrowExpectedRegExp = (
   expected: RegExp,
 ) => {
   const pass = thrown !== null && expected.test(thrown.message);
-  const isNotError = thrown !== null && !thrown.isError;
 
   const message = pass
     ? () =>
         matcherHint(matcherName, undefined, undefined, options) +
         '\n\n' +
         formatExpected('Expected pattern: ', expected) +
-        (isNotError
-          ? formatReceived('Received value:   ', thrown, 'value')
-          : formatReceived('Received message: ', thrown, 'message') +
-            formatStack(thrown))
+        (thrown !== null && thrown.hasMessage
+          ? formatReceived('Received message: ', thrown, 'message') +
+            formatStack(thrown)
+          : formatReceived('Received value:   ', thrown, 'value'))
     : () =>
         matcherHint(matcherName, undefined, undefined, options) +
         '\n\n' +
         formatExpected('Expected pattern: ', expected) +
         (thrown === null
           ? '\n' + DID_NOT_THROW
-          : isNotError
-          ? formatReceived('Received value:   ', thrown, 'value')
-          : formatReceived('Received message: ', thrown, 'message') +
-            formatStack(thrown));
+          : thrown.hasMessage
+          ? formatReceived('Received message: ', thrown, 'message') +
+            formatStack(thrown)
+          : formatReceived('Received value:   ', thrown, 'value'));
+
+  return {message, pass};
+};
+
+type AsymmetricMatcher = {
```

Comment:
Have you tried putting this here?
https://github.com/facebook/jest/blob/57aeaa21cef41755740e9a8698118fc7947e37a1/types/Matchers.js#L46
In case that causes type errors all over the place (there's probably a reason why it's `Object` 😄) we should probably save that costly refactoring for another time after Jest 24 is released, but maybe it works just fine?